src/neurostatslib/basis.py

```diff
@@ -533,6 +537,218 @@ def _check_n_basis_min(self) -> None:
+        The B-splines have (order-2) continuous derivatives at each interior knot.
+        The higher this number, the smoother the basis representation will be.
+
+    Attributes
+    ----------
+    order :
+        Spline order.
+
+    Methods
+    -------
+    _evaluate(x_tuple)
+       Evaluate the basis function at the samples x_tuple[0]. x_tuple must be of length 1 in order to pass the checks
+       of super().evaluate
+
+    References
+    ----------
+    ..[2] Prautzsch, H., Boehm, W., Paluszny, M. (2002). B-spline representation. In: Bézier and B-Spline Techniques.
+    Mathematics and Visualization. Springer, Berlin, Heidelberg. https://doi.org/10.1007/978-3-662-04919-8_5
+
+    """
+
+    def __init__(self, n_basis_funcs: int, order: int = 2):
+        super().__init__(n_basis_funcs, order=order)
+
+    def _evaluate(self, sample_pts: NDArray) -> NDArray:
+        """
+        Evaluate the B-spline basis functions with given sample points.
+
+        Parameters
+        ----------
+        sample_pts :
+            The sample points at which the B-spline is evaluated.
+
+        Returns
+        -------
+        NDArray
+            The basis function evaluated at the samples (Time points x number of basis)
+
+        Raises
+        ------
+        AssertionError
+            If the sample points are not within the B-spline knots range unless `outer_ok=True`.
+
+        Notes
+        -----
+        This method evaluates the B-spline basis functions at the given sample points.
+        It requires the knots to be defined through the `_generate_knots` method.
+
+        The evaluation is performed by looping over each element and using `splev`
+        from SciPy to compute the basis values.
+        """
+        super()._check_samples_non_empty(sample_pts)
+
+        # add knots
+        self._generate_knots(sample_pts, 0.0, 1.0)
+
+        basis_eval = bspline(
+            sample_pts, self.knot_locs, order=self.order, der=0, outer_ok=False
+        )
+
+        return basis_eval
+
+    def _check_n_basis_min(self) -> None:
+        """Check that the user required enough basis elements.
+
+        Check that BSplineBasis has at least as many basis as the order of the spline.
+
+        Raises
+        ------
+        ValueError
+            If an insufficient number of basis element is requested for the basis type
+        """
+        if self.n_basis_funcs < self.order:
+            raise ValueError(
+                f"{self.__class__.__name__} `order` parameter cannot be larger "
+                "than `n_basis_funcs` parameter."
+            )
+
+
+class CyclicBSplineBasis(SplineBasis):
+    """
+    B-spline 1-dimensional basis functions for cyclic splines.
+
+    Parameters
+    ----------
+    n_basis_funcs :
+        Number of basis functions.
+    order :
+        Order of the splines used in basis functions. Must lie within [1, n_basis_funcs].
```

Comment:
this isn't true for the cyclic one, is it?

Reply:
I figured it is actually true. I was wrapping too many knots in my original algorithm. that did not change the final results because I was adding a bunch of zero where the wrapping was not necessary.